### ISSUE TITLE
Automatic reconnect

### ISSUE THREAD

**Issue by jakubkulhan** [jakubkulhan/bunny#29]
- [ ] `Client` option to allow reconnect (+ reconnect timeout)
- [ ] open all previously opened channels
- [ ] start all previously started consumers
  - what about anonymous queues?
- [ ] return callbacks
- [ ] confirm callbacks
- [ ] MAYBE: declare all previously declared queues
  - only anonymous queues?
- [ ] MAYBE: declare all previously declared exchanges

**Comment by Vinceveve** [comment]
The main problem with automatic reconnect is that exceptions are thrown directly (no link with the promise). 
The usual errors (during inactivity or network latencies) are then thrown by the event loop [https://github.com/jakubkulhan/bunny/blob/master/src/Bunny/AbstractClient.php#L282].
The only way to catch error is to try catch the eventloop (not optimal)

How can I help to extract the broken pipe and link it with the produce or consume promise ?

**Comment by jakubkulhan** [comment]
Yes, rejecting pending promises should also be performed, however, it won't suffice alone.

**Comment by Vinceveve** [comment]
Here is a POC (with some reactiveX flavour) https://github.com/Domraider/bunny/blob/master/README.md 

Main problem, we have to create a channel for each publish

**Comment by cboden** [comment]
+1 on using Rx to solve this problem

**Comment by WyriHaximus** [comment]
Kinda hoping I can add this for 0.7 or 0.8. Does (n)ack messages work on a new connection if it dropped in the middle of handling it?
